build/FindDuplicateFiles.cs

```diff
@@ -0,0 +1,311 @@
+                Files.Length,
+                infos.Count,
+                Files.Length - infos.Count);
+
+            if (infos.Count == 0)
+            {
+                UniqueFiles = Array.Empty<ITaskItem>();
+                DuplicateFiles = Array.Empty<ITaskItem>();
+                return true;
+            }
+
+            var filePathToDuplicates = FindDuplicates(infos);
+
+            var fileCount = filePathToDuplicates.Sum(x => x.Value.Count);
+            var uniqueCount = filePathToDuplicates.Count;
+            var duplicateCount = fileCount - uniqueCount;
+            Log.LogMessage(
+                MessageImportance.High,
+                "Of the {0} unique file paths provided, {1} ({2:P}) are unique and {3} ({4:P}) are duplicate.",
+                fileCount,
+                uniqueCount,
+                (float)uniqueCount / fileCount,
+                duplicateCount,
+                (float)duplicateCount / fileCount);
+
+            var uniqueFiles = new List<ITaskItem>();
+            var duplicateFiles = new List<ITaskItem>();
+            foreach (var pair in filePathToDuplicates)
+            {
+                foreach (var duplicate in pair.Value)
+                {
+                    if (pair.Key == duplicate.FullPath)
+                    {
+                        uniqueFiles.Add(duplicate.Item);
+                    }
+                    else
+                    {
+                        duplicate.Item.SetMetadata("DuplicateOf", pair.Key);
+                        duplicateFiles.Add(duplicate.Item);
+                    }
+                }
+            }
+
+            UniqueFiles = uniqueFiles.ToArray();
+            DuplicateFiles = duplicateFiles.ToArray();
+
+            return true;
+        }
+
+        private List<TaskItemInfo> GetUniqueTaskItemInfo()
+        {
+            // Compare paths in a case insensitive manner.
+            var uniqueFullPaths = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
+
+            var infos = new List<TaskItemInfo>();
+            foreach (var item in Files)
+            {
+                var fullPath = item.GetMetadata("FullPath");
+                if (fullPath == null || !File.Exists(fullPath))
+                {
+                    Log.LogWarning("File '{0}' does not exist.", item.ItemSpec);
+                    continue;
+                }
+
+                if (!uniqueFullPaths.Add(fullPath))
+                {
+                    Log.LogWarning("File path '{0}' is associated with multiple items. Only the first will be used.", item.ItemSpec);
+                    continue;
+                }
+
+                var info = new TaskItemInfo(item, fullPath);
+
+                infos.Add(info);
+            }
+
+            return infos;
+        }
+
+        private Dictionary<string, List<TaskItemInfo>> FindDuplicates(List<TaskItemInfo> infos)
+        {
+            var fileSizeToInfo = GetFileSizeToInfo(infos);
```

Comment:
Super nit: pluralize infos to match the `infos` parameter?

```suggestion
            var fileSizeToInfos = GetFileSizeToInfos(infos);
```